skimage/feature/corner_cy.pyx

```diff
@@ -5,13 +5,258 @@
 import numpy as np
 cimport numpy as cnp
 from libc.float cimport DBL_MAX
-from libc.math cimport atan2, fabs
+from libc.math cimport M_PI, atan, atan2, cos, fabs, fmax, sqrt
 
 from .._shared.fused_numerics cimport np_floats
 from ..util import img_as_float64
 
 cnp.import_array()
 
+def _real_symmetric_2x2_evs(
+    np_floats[:, ::1] M00, np_floats[:, ::1] M01, np_floats[:, ::1] M11,
+    bint ascending=False, bint abs_sort=False,
+):
+    """Analytical eigenvalues of a symmetric 2 x 2 matrix."""
+    # use double-precision intermediate variables for accuracy
+    cdef:
+        cnp.float64_t m00, m01, m11, tmp1, tmp2, lam1, lam2, stmp;
+        Py_ssize_t rows = M00.shape[0];
+        Py_ssize_t cols = M00.shape[1];
+        Py_ssize_t r, c;
+
+    if np_floats is cnp.float32_t:
+        dtype = np.float32
+    else:
+        dtype = np.float64
+
+    cdef np_floats[:, :, ::1] evals = np.empty((2, rows, cols), dtype=dtype)
+
+    with nogil:
+        for r in range(rows):
+            for c in range(cols):
+                m00 = <cnp.float64_t>(M00[r, c])
+                m01 = <cnp.float64_t>(M01[r, c])
+                m11 = <cnp.float64_t>(M11[r, c])
+                tmp1 = 4 * m01 * m01
+
+                tmp2 = m00 - m11
+                tmp2 *= tmp2
+                tmp2 += tmp1
+                tmp2 = sqrt(tmp2) / 2
+
+                tmp1 = (m00 + m11) / 2
+                if ascending:
+                    lam1 = tmp1 - tmp2
+                    lam2 = tmp1 + tmp2
+                    if abs_sort and (fabs(lam1) > fabs(lam2)):
+                        stmp = lam1
+                        lam1 = lam2
+                        lam2 = stmp
+                else:
+                    lam1 = tmp1 + tmp2
+                    lam2 = tmp1 - tmp2
+                    if abs_sort and (fabs(lam1) < fabs(lam2)):
+                        stmp = lam1
+                        lam1 = lam2
+                        lam2 = stmp
+                evals[0, r, c] = lam1
+                evals[1, r, c] = lam2
+    return np.asarray(evals)
+
+
+def _real_symmetric_3x3_evs(
+    np_floats[:, :, ::1] M00, np_floats[:, :, ::1] M01, np_floats[:, :, ::1] M02,
+    np_floats[:, :, ::1] M11, np_floats[:, :, ::1] M12, np_floats[:, :, ::1] M22,
+    bint ascending=False, bint abs_sort=False,
+):
+    """Analytical eigenvalues of a symmetric 3 x 3 matrix.
+
+    Follows the expressions given for hermitian symmetric 3 x 3 matrices in
+    [1]_, but simplified to handle real-valued matrices only.
+
+    Parameters
+    ----------
+    a, d, f, b, e, c : cp.ndarray
+        Images corresponding to the individual components of the matrix, `M`,
```

Comment:
Where above is M shown?

Reply:
I had moved it into the code comments in the inner loop below, but missed updating the docstring here. These should say `M00, M01, M02, M11, M12, M22` instead of `a, d, f, b, e, c`. Thanks for catching that.